### ISSUE TITLE
Replace proprietary Google Fused location library with open source microG library

### ISSUE THREAD

**Issue by CoelacanthusHex** [breezy-weather/breezy-weather#901]
### Describe your idea

Now fused location feature is only enabled in Google Play flavor. But it can be implemented with open source projects and services completely. To use this feature, we must include a library and a runtime application. Both of them can be implemented with OSS easily. For the former, just replace https://github.com/breezy-weather/breezy-weather/blob/main/app/build.gradle.kts#L217 with OSS implementation https://github.com/microg/GmsCore/tree/master/play-services-location, which is compatible with gms API by API but completely open source, even doesn't use non-free network service. Some open source projects in F-droid have started to use microG to implemented these useful feature in easy and open source way, such as Fennec, which [uses microG to implement FIDO feature](https://gitlab.com/relan/fennecbuild/-/blob/master/gecko-liberate.patch). For the latter, just use microG to replace Google's GMS.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] My idea is not related to a question covered in the [Frequently Asked Questions](https://github.com/breezy-weather/breezy-weather/blob/main/HELP.md) document, or a feature described in the [Homepage](https://github.com/breezy-weather/breezy-weather/blob/main/docs/HOMEPAGE.md) document.
- [X] I have updated the app to **[version 5.1.8 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
> doesn't use non-free network service

How is that possible?

**Comment by CoelacanthusHex** [comment]
> > doesn't use non-free network service
> 
> How is that possible? If the server-side is also open source, it's not Google's, it's an alternative service, right?

Yes. And there is a local database as well, which records WiFi and Cell when GPS works and use it to determine location later.

**Comment by papjul** [comment]
So, we can't replace it or users expecting Google's accuracy won't have it anymore?

Additionally I don't plan on bundling a 2 GB local database, requesting additional permissions nearby wifi and bluetooth and more, it's overkill.

**Comment by CoelacanthusHex** [comment]
> So, we can't replace it or users expecting Google's accuracy won't have it anymore?

With open source network service, the accuracy is worse than Google's. With local database, it could be much better than Google's where you have reached.

> 
> Additionally I don't plan on bundling a 2 GB local database, requesting additional permissions nearby wifi and bluetooth and more, it's overkill.

No. There is no additional database when build. microG just collect data and build database on user's devices locally when they use the device. And Breezy Weather needn't any additional permission: Breezy Weather request location from microG application installed on the same device. microG will do all the thing, Breezy Weather just need call a API.

**Comment by papjul** [comment]
So, that's a completely different location provider? It needs a separate entry in Settings > Location > Location service, then.

If the user needs to install microG Services, it's the same thing as if they need to install Google Play Services. It won't work on devices without it.

I will leave this open to contributions.

**Comment by CoelacanthusHex** [comment]
PS: microG and GMS are both service and library, and they can be combined in any combination:
- compile with microG library and use with microG service
- compile with GMS library and use with GMS service
- compile with microG library and use with GMS service
- compile with GMS library and use with microG service

**Comment by CoelacanthusHex** [comment]
> > microG and GMS are both service and library, and they can be combined in any combination
> 
> That's much clearer now, thanks!
> 
> > In my opinion, I want to push this source into Standard and F-Droid flavors.
> 
> But there are also users like me without any *Services that must still be able to use the current location without it failing.
> 
> So, at the moment, the idea is to replace the library of the `gplay` flavor and have someone with Google Play Services test it still works.
> 
> To use it in `standard`, we would need to add a condition "IF google play services are installed OR microG services are installed OR Huawei services are installed OR whatever services are installed THEN use Google Fused location" but that's not very flexible. I saw that there is an API to check the availability of Google Play Services, I believe microG would say `true` here? Maybe we could use that. Is there a documentation of that microG library so I can understand better what to expect?

microG is designed to have the same behavior as GMS except non-privacy behaviors. There are few documents about microG for developers because it's the same as GMS in most cases. If you have any questions, @mar-v-in and @ale5000-git may give you some answers. And @relan has some experience using microG to replace GMS in Fennec (A full free Firefox build).

> 
> I also don't think we can really qualify the Google Fused location as a "free network" because someone reversed engineered, spoofed Google signature and re-used the same API. It's still a non-free network in the way it was intended and for the majority of users who have Google Play Services and not microG. So adding it to `standard` why not since there would no longer be any proprietary lib bundled, but `fdroid` I would like to avoid as this is a no "non-free network" flavor. F-Droid users can still use the basic network location provider from Google Play Services if they wish to, but that's something we can't prevent.

Ok, at least this source will have a place in Standard flavor. For the non-free network, @linsui from F-droid may have some opinions.

**Comment by linsui** [comment]
In F-Droid flavor as long as the fused location is not used by default it's not considered as NonFreeNet.

**Comment by papjul** [comment]
WIP on https://github.com/breezy-weather/breezy-weather/tree/microg branch if interested

I don't have Google Play Services or microG so feel free to test

**Comment by CoelacanthusHex** [comment]
> WIP on [`microg`](https://github.com/breezy-weather/breezy-weather/tree/microg?rgh-link-date=2024-04-08T20%3A21%3A48Z) branch as a separate location source you can select from `Settings` > `Location` if interested
> 
> I don't have Google Play Services or microG so feel free to test

I can confirm that the Fused Location works well with microG 0.3.1.

**Comment by CoelacanthusHex** [comment]
> WIP on [`microg`](https://github.com/breezy-weather/breezy-weather/tree/microg?rgh-link-date=2024-04-08T20%3A21%3A48Z) branch as a separate location source you can select from `Settings` > `Location` if interested
> 
> I don't have Google Play Services or microG so feel free to test

@Cryolitia and @KunoiSayami confirmed it works with GMS.

**Comment by Cryolitia** [comment]
Confirmed it works on Sony Xperia 1 IV, firmware version 64.2.A.2.112.

Tested-by: Cryolitia <Cryolitia@gmail.com>

**Comment by papjul** [comment]
Currently, the Google Fused Location uses both Fused + Android framework in parallel, so if Fused fails, you still have data.

So, I removed Android framework calls to keep only the efficient Fused.

You can download the debug artifact from: https://github.com/breezy-weather/breezy-weather/pull/914/checks

Please test it that way:
1. Tap the back button to exit the (empty) location list
2. Tap the Settings icon
3. Go to Location settings
4. Choose "Google Fused" as the location service
5. Go back to location list
6. Add your current location

If it works, let me know! If it doesn't work, please export the crash log from `Settings` > `Debug`

Thanks!

**Comment by Cod3dDOT** [comment]
FYI, microg does not support the recommended way to get the current location:
This is what 3/4 recommended methods look like. 

https://github.com/microg/GmsCore/blob/master/play-services-location/src/main/java/org/microg/gms/location/FusedLocationProviderClientImpl.java
```
@NonNull
@Override
public Task<Location> getCurrentLocation(int priority, CancellationToken cancellationToken) {
        return null;
}

@NonNull
@Override
public Task<Location> getCurrentLocation(CurrentLocationRequest request, CancellationToken cancellationToken) {
        return null;
}

@NonNull
@Override
public Task<Location> getLastLocation(LastLocationRequest request) {
        return null;
}

public Task<Location> getLastLocation() {
        return scheduleTask((ReturningGoogleApiCall<Location, LocationClientImpl>) LocationClientImpl::getLastLocation);
}
```

**Comment by CoelacanthusHex** [comment]
> As far as I can tell, microg does not support the recommended way to get the current location ([getCurrentLocation](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#getCurrentLocation(com.google.android.gms.location.CurrentLocationRequest,%20com.google.android.gms.tasks.CancellationToken))).
> 
> See [microg FusedLocationProviderClientImpl](https://github.com/microg/GmsCore/blob/master/play-services-location/src/main/java/org/microg/gms/location/FusedLocationProviderClientImpl.java?rgh-link-date=2024-04-10T01%3A46%3A19Z#L33-L43)
> 
> The app currently uses [requestLocationUpdates](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient#requestLocationUpdates(com.google.android.gms.location.LocationRequest,%20java.util.concurrent.Executor,%20com.google.android.gms.location.LocationListener)), which is supported, but intended for tracking rather than locating.

I think it's not a problem in current way: it only affect users use BW with microG service, it doesn't affect users use BW compiled with microG but with GMS service.

Of course, it's better to let upstream support other methods.

**Comment by CoelacanthusHex** [comment]
> If it works, let me know! If it doesn't work, please export the crash log from `Settings` > `Debug`

I can confirm it works well with microG 0.3.1.

**Comment by papjul** [comment]
We have another important bug that makes it a no-go for now: see microg/GmsCore#2309 

We will keep the Google Play lib on `gplay` flavor for now, but we will implement Google Fused as a separate entry from Android source

**Comment by papjul** [comment]
There is a 3rd option that I randomly found and is barely used in any other app (because Android framework documentation for location services is nearly inexistant): Fused provider is natively supported on Android >= 12 and `QUALITY_BALANCED_POWER_ACCURACY` is available on all versions through a Compat class.

Basically, this means we can decide to not depend on any library, and use the native Android functions, but this will only supports Android >= 12 (previous versions will just use the regular Android source, both network and GPS). Given all Android versions < 12 are EoL, I think it's an acceptable trade-off to get rid of a dependency and a flavor.

WIP on `nativefused` branch

**Comment by papjul** [comment]
This branch was tested on different phones successfully. This makes this request obsolete as we no longer depend on any library
